### DISCUSSION TITLE
Discussion: Consider making a "shortcut" API around TypeScript for node types

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6013] [RFCs]
### Suggestion

We write code like this very frequently:

```ts
const parserServices = util.getParserServices(context);
const checker = parserServices.program.getTypeChecker();
const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
const type = checker.getTypeAtLocation(node);
```

It's a little burdensome.

@bradzacher and I were chatting in a pairing about how it might be convenient to make a wrapper around this, for convenience. Vague first draft:

```ts
declare const typeUtils: {
  [K in keyof ts.TypeChecker]:
    (services: ParserServices, node: TSESTree.Node) => ReturnType<ts.TypeChecker[K]>;
};

const type = typeUtils.getTypeAtLocation(parserServices, node);
```

We'd probably want to export this as a publicly available & documented API, so it's not confusing to consumers why our source code doesn't look like examples.

One downside would be that we're obfuscating how to use TypeScript APIs. I hate to add to complexity of understanding these already hard-to-understand things...

**Comment by bradzacher** [comment]
I would say we could probably go one step further and just offer all of this off of the parser serivces!

For example:
```ts
type ParserServices = {
  getTypeAtLocation(node: TSESTree.Node): ReturnType<ts.TypeChecker['getTypeAtLocation']>;
};

const parserServices = util.getParserServices(context);

parserServices.getTypeAtLocation(node);
```

That would make it even more straight-forward for people to use the tooling!

**Comment by JoshuaKGoldberg** [comment]
Oh interesting. I like putting this as a member of `ParserServices`, though I'm hesitant to put potentially dozens of them as siblings of our own properties. Maybe `parserServices.checker.getTypeAtLocation`?

**Comment by bradzacher** [comment]
the reason I like it as a direct sibling is it's one less step.
Right now people do things like
```ts
const {program} = util.getParserServices(context);
const checker = program.getTypeChecker();
```
Adding it as a separate property would make it more likely there's a clash if the consumer needs a more advanced API than what we offer.
For example:
```ts
const {checker, program} = util.getParserServices(context);
const checker2 = program.getTypeChecker();
```

If we build it right then most usecases shouldn't really need to ever manually touch `program`/`esTreeNodeToTSNodeMap`/`tsNodeToESTreeNodeMap`!
